### PR TITLE
Share Extension Part Three

### PR DESCRIPTION
**Pull request by aerych**
Refs #91 

This PR is part three of a four part series adding Share Extension support for sharing images to Newspack.  In this PR we do the following: 
- Reorganizes more code into the NewspackFramework to share between the extension and main app without duplication. In this case, ColorStudio and associated classes.
- Introduces a custom UI for the share extension.
- Wrangles shared items, and persists shared items to group storage on save. NSItemProvider wrangling is courtesy of code borrowed from WPiOS that has been slightly modified for Newspack.

This is another PR where a large portion of the changes are boiler plate, or due to incorporating existing production code from another app. New changes that need a keen eye are in: 
- FileManager+Helpers.swift (improves file suffix handling)
- SharedMediaViewController.swift
- StorySelectorViewController.swift
- ShareAppearance.swift

The following files are slightly modified from their WPiOS counterparts, mostly to remove code that is not relevant to Newspack at this time. A passing glance wouldn't hurt but willing to trust the previous implementation.
- ShareExtractor.swift
- UIImage_ShareExtractor.swift
- NSExtensionContext+ShareExtractor.swift

#### Scenario 1:  Logged out
Prep: Be logged out of Newspack.  Launch the share extension with the Photos app.
- [x]  Confirm a prompt appears that sharing isn't possible, and asks you to log in to Newspack before trying again. 
- [x]  Confirm clicking the prompt's OK button dismisses the share extension and you're returned to the Photos share screen.

#### Scenario 2: Logged in
Prep: Be logged in to Newspack with several stories. Launch the share extension with the Photos app.
- [x] Confirm you do not see the warning prompt.
- [x] Confirm the UI looks good in portrait and landscape orientation, and in dark and light modes.
- [x] Confirm you see the photos you selected as thumbnails.
- [x] Confirm the selected story in the share extension is the currently selected story in the main app.
- [x] Tap the story cell and use the following screen to change the story that will be shared to.
- [x] Confirm tapping the cancel button dismisses the share extension and you're returned to the Photos share screen.
- [x] Confirm tapping the save button dismisses the share extension and you're returned to the Photos app (exit all the way out of sharing). 

#### Scenario 3: Confirm cast shadows
- [x] Launch the main app with the debugger.  Observe what is logged to the console.  Confirm you see the total items shared logged as shadow assets.

#### Scenario 4:  
- [x] Launch the share extension on the iPad.  Confirm it appears as a large modal form sheet and the layout and style looks good in portrait and landscape orientations, and light and dark mode. 

### Example Screens:
Simulator Screen Shot - iPhone SE (2nd generation) - 2020-09-04 at 10 15 46
Simulator Screen Shot - iPhone SE (2nd generation) - 2020-09-04 at 10 15 53
Simulator Screen Shot - iPhone SE (2nd generation) - 2020-09-04 at 10 15 58
Simulator Screen Shot - iPad Pro (9 7-inch) - 2020-09-04 at 10 14 09

### CODE REVIEW

File: Newspack/NewspackFramework/Extensions/FileManager+Helpers.swift [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        return url
+    }
+
+    private func appendSuffix(suffix: String, to path: String) -> String {
+        guard path.contains(".") else {
+            return path + suffix
+        }
+
+        var parts = path.components(separatedBy: ".")
+        let str = parts[parts.count - 2] + suffix
```
Comment (jleandroperez) [inline Newspack/NewspackFramework/Extensions/FileManager+Helpers.swift:28]: Is it possible that `parts` could ever not have a dot?

File: Newspack/NewspackShare/Controllers/ShareMediaViewController.swift [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+        let action = UIAlertAction(title: actionTitle, style: .default, handler: { [weak self] _ in
+            self?.cancelSharing()
+        })
+        alert.addAction(action)
+
+        present(alert, animated: true, completion: nil)
+    }
+
+    func cancelSharing() {
+        let error = NSError(domain: "com.auttomattic.newspack.share", code: 0, userInfo: nil)
```
Comment (jleandroperez) [inline Newspack/NewspackShare/Controllers/ShareMediaViewController.swift]: Would the domain be worthy of a constant?

File: Newspack/NewspackShare/Controllers/ShareMediaViewController.swift [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+        cell.textLabel?.text = targetStory?.title
+        cell.detailTextLabel?.text = NSLocalizedString("Change", comment: "Verb. An action. Indicates that a selection can be changed via an interaction.")
+        cell.accessoryType = .disclosureIndicator
+
+        return cell
+    }
+
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard let controller = UIStoryboard.init(name: "MainInterface", bundle: nil).instantiateViewController(withIdentifier: "StorySelectorViewController") as? StorySelectorViewController else {
```
Comment (jleandroperez) [inline Newspack/NewspackShare/Controllers/ShareMediaViewController.swift]: Nipticky: `.init` is not needed 😄

File: Newspack/NewspackShare/Utils/ShareExtractor.swift [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+private protocol ExtensionContentExtractor {
+    func canHandle(context: NSExtensionContext) -> Bool
+    func extract(context: NSExtensionContext, completion: @escaping ([ExtractedItem]) -> Void)
+    func saveToSharedContainer(image: UIImage) -> URL?
+    func saveToSharedContainer(wrapper: FileWrapper) -> URL?
+    func copyToSharedContainer(url: URL) -> URL?
+}
+
+private protocol TypeBasedExtensionContentExtractor: ExtensionContentExtractor {
+    associatedtype Payload
```
Comment (jleandroperez) [inline Newspack/NewspackShare/Utils/ShareExtractor.swift:98]: Nice!!!

File: Newspack/NewspackShare/Controllers/ShareMediaViewController.swift [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    ///
+    func validateData() -> Bool {
+        guard
+            shadowSites.count > 0,
+            let _ = targetStory
+        else {
+            return false
+        }
+
+        return true
```
Comment (jleandroperez) [inline Newspack/NewspackShare/Controllers/ShareMediaViewController.swift]: Nipticky!! (Feel super free to disregard!!)

```
return shadowSites.count > 0 && targetStory != nil
```

File: Newspack/NewspackShare/Controllers/ShareMediaViewController.swift (hunk 2 above)
Comment (jleandroperez) [reply Newspack/NewspackShare/Controllers/ShareMediaViewController.swift]: Update: Probably not!!

**Review by jleandroperez — APPROVED** [review]
All of the test scenarios look good!! A couple nipticky code Q's, aaand another note:

```
Tap the story cell and use the following screen to change the story that will be shared to.
```
 
In this test scenario, a potential good enhancement would be to make the  "Story Picker" ViewController highlight the currently selected Story.

Other than that, :shipit: sir!!!

File: Newspack/NewspackShare/Controllers/ShareMediaViewController.swift (hunk 3 above)
Comment (aerych) [reply Newspack/NewspackShare/Controllers/ShareMediaViewController.swift]: I don't know why I sometimes do that 🤦

File: Newspack/NewspackShare/Controllers/ShareMediaViewController.swift (hunk 2 above)
Comment (aerych) [reply Newspack/NewspackShare/Controllers/ShareMediaViewController.swift]: No no... good call actually.  As more types are supported by the share extension there may be more opportunities for dispatching/logging NSErrors.

File: Newspack/NewspackShare/Controllers/ShareMediaViewController.swift (hunk 5 above)
Comment (aerych) [reply Newspack/NewspackShare/Controllers/ShareMediaViewController.swift]: :)  I like the longer form for readability at a glance.   Good suggestion tho! Keep 'em coming!

Comment (aerych) [reply Newspack/NewspackShare/Controllers/ShareMediaViewController.swift]: Hmm... Actually after a quick second thought... I like you're suggestion better!  Call me fickle 😆

File: Newspack/NewspackFramework/Extensions/FileManager+Helpers.swift (hunk 1 above)
Comment (aerych) [reply Newspack/NewspackFramework/Extensions/FileManager+Helpers.swift:28]: Possibly!  But that scenario should be handled by the guard statement above.

**Comment by aerych** [comment]
> a potential good enhancement would be to make the "Story Picker" ViewController highlight the currently selected Story

Excellent idea.

**Comment by aerych** [comment]
Thanks @jleandroperez !  Changes made.  One last quick peek for paranoia sake? 

Screen shot showing the selected cell:
Simulator Screen Shot - iPhone SE (2nd generation) - 2020-09-04 at 13 03 36
It's a little different from the main app, but the main app is probably going to change too so this is a good placeholder.

**Review by jleandroperez — APPROVED** [review]
Re-tested all the things, looks great sir!!

:shipit:

**Comment by aerych** [comment]
Thank you @jleandroperez !!

**Pull request merged**